ci/ipsec: Cilium agents in ci-ipsec-e2e no longer share host's boot ID

jschwinger233 opened the pull request:
Containers on the same host share boot id (/proc/sys/kernel/random/boot_id), previous IPsec kind-based e2e tests didn't fully cover the per-node-pair key feature.

This PR introduces a new hidden flag `boot-id-file` for specifying boot ID path, allows us to cover the missing scenario in CI.

ci-ipsec-upgrade can't be covered in this PR because 1.16 still shares boot ID. Upgrade from 1.16 to 1.17 in CI will break connectivity as new cilium suddenly uses a different boot ID.

pchaigno commented on .github/workflows/conformance-ipsec-e2e.yaml:
```diff
@@ -164,7 +164,7 @@ jobs:
         uses: ./.github/actions/cilium-config
         with:
-          image-tag: ${{ steps.vars.outputs.sha }}
+          image-tag: ${{ steps.vars.outputs.sha }}-bootid-override
```
Don't we need the same for the upgrade workflow?

jschwinger233 replied:
This implementation of per-kind-node bootid is actually per-agent-process bootid, which changes after agent restarts (upgrade). ci-ipsec-upgrade goes red due to that.

I'll add a hidden flag since everyone is okay with it. Hidden flag solution is more low-carbon (doesn't need to run a new ci build job) and it can be real per-kind-node bootid.

pchaigno reviewed (approved):
I don't mind this approach, but I'm wondering why you went this way rather than adding a hidden agent flag? AFAICS, there's no risk of silently breaking security here. If some user manages to use the hidden flag, they'll just break connectivity, as is already possible with many other flags.

jschwinger233 committed:
Add a hidden flag to specify boot ID filename

By default, Cilium agents read the boot ID from
/proc/sys/kernel/random/boot_id, which is shared among containers
running on the same host. This commit introduces a hidden flag for
specifying the boot ID filename. With this flag set to a different
location, containerized Cilium agents can avoid sharing the same boot
ID.

The previous implementation read /proc/sys/kernel/random/boot_id during
init(), which is no longer feasible because we now need to wait until
the daemon processes the flags. To address this, an initLocalBootID()
function has been introduced to replace init(). Additionally, a new
file, bootid_other.go, has been created to handle non-Linux systems.

Signed-off-by: gray <gray.liang@isovalent.com>

jschwinger233 committed:
ci-ipsec-e2e: agents in local kind don't share boot IDs

/var/run/cilium/ is mounted to cilium-agent pod via hostPath:

  volumeMounts:
  - mountPath: /var/run/cilium
    name: cilium-run
  volumes:
  - hostPath:
      path: /var/run/cilium
      type: DirectoryOrCreate
    name: cilium-run

By configuring /var/run/cilium/boot_id with a random uuid in Kind
containers, Cilium agents installed with the hidden flag
`boot-id-file=/var/run/cilium/boot_id` will use different boot IDs.

This fixes the CI issue that Cilium agents running locally in Kind
always shared the same boot ID.

Signed-off-by: gray <gray.liang@isovalent.com>
